### PR TITLE
Use DefaultArgumentExpr for caller-side defaults

### PR DESCRIPTION
**Pull request by hamishknight**
This PR changes how we represent caller-side default arguments within the AST. Instead of directly inserting them into the call-site, use a DefaultArgumentExpr to refer to them indirectly.

The main goal of this change is to make it such that the expression type-checker no longer cares about the difference between caller-side and callee-side default arguments. In particular, it no longer cares about whether a caller-side default argument is well-formed when type-checking an apply. This is important because any conversions introduced by the default argument shouldn't affect the score of the resulting solution.

Instead, caller-side defaults are now lazily type-checked when we want to emit them in SILGen. This is done through introducing a request, and adjusting the logic in SILGen to be more lenient with ErrorExprs. Caller-side defaults in primary files are still also currently checked as a part of the declaration by `checkDefaultArguments`.

Resolves SR-11085.
Resolves rdar://problem/56144412.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckConstraints.cpp [hunk 1]
```diff
@@ -2323,12 +2323,12 @@ Type TypeChecker::typeCheckParameterDefault(Expr *&defaultValue,
...
     return typeCheckExpression(defaultValue, DC,
                                TypeLoc::withoutLoc(fnType->getResult()),
-                               canFail ? CTP_DefaultParameter : CTP_CannotFail,
-                               TypeCheckExprOptions(), &listener);
+                               CTP_DefaultParameter, TypeCheckExprOptions(),
+                               &listener);
   }
 
   return typeCheckExpression(defaultValue, DC, TypeLoc::withoutLoc(paramType),
-                             canFail ? CTP_DefaultParameter : CTP_CannotFail);
```
Comment (xedin) [inline lib/Sema/TypeCheckConstraints.cpp]: Is it possible to remove `CTP_CannotFail` now?

**Review by xedin — APPROVED** [review]
Looks great, thank you!

File: lib/Sema/TypeCheckConstraints.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/Sema/TypeCheckConstraints.cpp]: @xedin Unfortunately it's still used in a couple of places in CSApply :/ Thanks for the review!

Comment (xedin) [reply lib/Sema/TypeCheckConstraints.cpp]: No worries, soon then :)

**Comment by hamishknight** [comment]
@swift-ci please test

**Pull request merged**